From: Martin Schwidefsky <schwidefsky@de.ibm.com>
Subject: s390/3270: fix garbled output on 3270 tty view
Patch-mainline: v4.7-rc1
Git-commit: 2e63a3a66655d5fe5d85c090b009979870436c00
References: bnc#980892, LTC#141736

Description:  kernel: fix disconnect/reconnect of 3270 terminals
Symptom:      The disconnect and reconnect of a 3270 terminal is not
              working as expected, the terminal session is not hung
              up and the new size of the reconnected terminal is not
              detected.
Problem:      The 3270 driver does not have support for disconnected
              terminals.
Solution:     Add support for disconnected 3270 terminals.
Reproduction: Open a 3270 terminal, login on the tty, disconnected the
              terminal and reconnect with a terminal of different size.

Upstream-Description:

              s390/3270: fix garbled output on 3270 tty view

              The tty3270_update function tries to optimize the output stream by
              skipping the TO_SBA command to position the output offset if the
              TO_RA command of the previous line ended at the same offset.
              But some lines do not have the TO_RA at the end, namely those
              with a length equal to the number of columns.

              Make sure there is a TO_RA at the end of the previous line to avoid
              funny looking output.

              Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>


Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 drivers/s390/char/tty3270.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

--- a/drivers/s390/char/tty3270.c
+++ b/drivers/s390/char/tty3270.c
@@ -405,7 +405,10 @@ tty3270_update(struct tty3270 *tp)
 			if (raw3270_request_add_data(wrq, str, len) != 0)
 				break;
 			list_del_init(&s->update);
-			sba = s->string + s->len - 3;
+			if (s->string[s->len - 4] == TO_RA)
+				sba = s->string + s->len - 3;
+			else
+				sba = invalid_sba;
 		}
 		if (list_empty(&tp->update))
 			updated |= TTY_UPDATE_LIST;
